src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs

```diff
@@ -7513,4 +7513,333 @@ static class Interceptors
         var method = model.GetInterceptorMethod(node);
         Assert.Equal($"void Interceptors.M1(this {refKind}S s)", method.ToTestDisplayString());
     }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71657")]
+    public void ReceiverCapturedToTemp_StructRvalueReceiver()
+    {
+        var source = CSharpTestSource.Parse("""
+            using System;
+
+            public struct S
+            {
+                void M() => Console.WriteLine(0);
+
+                public static void Main()
+                {
+                    new S().M();
+                }
+            }
+            """, "Program.cs", RegularWithInterceptors);
+
+        var comp = CreateCompilation(source);
+        var model = comp.GetSemanticModel(source);
+        var node = source.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Last();
+        var locationSpecifier = model.GetInterceptableLocation(node)!;
+
+        var interceptors = CSharpTestSource.Parse($$"""
+            using System;
+
+            public static class C
+            {
+                {{locationSpecifier.GetInterceptsLocationAttributeSyntax()}}
+                public static void M1(this ref S s) => Console.WriteLine(1);
```

Comment:
Consider using `s` in a way that mutates, to show that we are actually capturing by ref and that the results are observable. The same for the rest of the test methods.

Reply:
I am wondering how exactly should I demonstrate this? Do you want me to try and read the managed reference at the call site (e.g. by returning 'ref S s' by-reference)?

Reply:
Yeah, that would work.

Reply:
> read the field from the local

The lowering temp which is being passed by-ref to the call can't be directly used in user code. We would have to return a ref to that temp if we wanted the call site to be able to read it.

Reply:
Am I misunderstanding the original bug? I expect something like this would be accessible, and would have triggered the emit assert:

```cs
var s = new S();
s.M(); // Intercepted, and `s` was mutated
```

Reply:
That case would not trigger the assert. The assert occurs when the compiler needs to create a temp for the receiver to use as an argument for a by-ref `this` parameter, such as `new S().M()`. It won't occur when the receiver argument is already something we can pass by-reference for the `this` parameter.

The emit layer is fine with creating a temp for an instance call receiver in the `new S().M()` case. However, when we intercept the call with a method like `static Extension(this ref S s)`, it basically sees the call as `Extension(new S())`, doesn't expect to create a temp for `new S()`, and will assert.

Reply:
Could we observe this with a `ref`-returning method? IE, something like this?

```cs
var s = new S();
Identity(ref s).M();
s.ObserveMutated();

ref S Identity(ref S s) => ref s;
```

Reply:
I think in this case the compiler also doesn't need to capture the receiver 'Identity(ref s)' to temp. But we can go ahead and add the test to show that we use the reference properly here.

Reply:
Think we could observe it this way 

```csharp
using System;
using System.Diagnostics.CodeAnalysis;

Span<int> Test() {
  // should error in both intercepted and non intercepted case
  return new S().M();
}

ref struct S {
    [UnscopedRef]
    public Span<int> M() => default;
}

static class D
{
    [InterceptsLocation(1, "...")]
    public static Span<int> Interceptor(this ref S s)
}
```